### ISSUE TITLE
wiring: App runner + Metrics + Scheduler + Makefile

### ISSUE THREAD

**Issue by skinnymanmusic** [Trialogic/TrialSync-ETL#29]
This issue provides a ready-to-paste wiring patch to run the ETL app locally with safe environment loading, preflight checks, metrics, and scheduler integration. It assumes you have merged the Dev/Test environment setup from Issue #25 (env loader + preflight), and the Week 2 scaffolds (#18 Scheduler, #19 CLI, #20 Metrics). If any of those are not merged yet, you can still paste these files and adjust imports accordingly.

---
## Files to add/update

1) src/app.py (new)
```python
import os
import signal
import sys
import time
from contextlib import contextmanager

try:
    from prometheus_client import start_http_server
except Exception:  # pragma: no cover
    start_http_server = None  # metrics optional

# Settings and preflight (from Issue #25)
from src.config.env import load_settings  # provides load_settings(env_path)
from src.config.preflight import assert_safe_nonprod  # blocks prod host in dev/test

# Core components
from src.api.client import ClinicalConductorClient
from src.db.loader import DataLoader
from src.etl.orchestrator import ETLOrchestrator

# Scheduler service (from Issue #18 scaffold)
try:
    from src.etl.scheduler import SchedulerService
except Exception:
    SchedulerService = None


def _get_env_file_from_env() -> str:
    # ENVIRONMENT can be: development|test|staging|production
    env = os.getenv("ENVIRONMENT", "development").strip()
    return f".env.{env}"


@contextmanager
def graceful_shutdown(orchestrator=None, scheduler=None):
    terminate = {"flag": False}

    def handler(signum, frame):
        terminate["flag"] = True

    signal.signal(signal.SIGINT, handler)
    signal.signal(signal.SIGTERM, handler)

    try:
        yield terminate
    finally:
        # Stop scheduler first to prevent new jobs from starting
        if scheduler is not None:
            try:
                scheduler.stop()
            except Exception:
                pass
        # Orchestrator-specific cleanup can be added here


def main():
    # 1) Load settings
    env_file = os.getenv("ENV_FILE", _get_env_file_from_env())
    settings = load_settings(env_file)

    # 2) Preflight safety for dev/test
    assert_safe_nonprod(settings)

    # 3) Metrics server (optional)
    if getattr(settings, "enable_metrics", False) and start_http_server is not None:
        port = int(getattr(settings, "metrics_port", 9090))
        start_http_server(port)
        print(f"[app] Metrics server started on :{port}")

    # 4) Construct core services
    api = ClinicalConductorClient(settings.cc_api_base_url, settings.cc_api_key)

    # NOTE: replace with your actual DB connection factory if different
    # For example, if you have a helper `get_db_connection(settings)` use it instead
    import psycopg2
    conn = psycopg2.connect(
        host=settings.database_host,
        port=settings.database_port,
        dbname=settings.database_name,
        user=settings.database_user,
        password=settings.database_password,
    )

    loader = DataLoader(conn)
    orchestrator = ETLOrchestrator(db_connection=conn, api_client=api, data_loader=loader)

    # 5) Scheduler service (optional)
    scheduler = None
    if SchedulerService is not None and getattr(settings, "enable_scheduler", True):
        scheduler = SchedulerService(orchestrator=orchestrator)
        # Option A: load schedules from DB table (dw_etl_schedules) if implemented
        try:
            scheduler.refresh_from_database()
            scheduler.start()
            print("[app] Scheduler started")
        except Exception as e:
            print(f"[app] Scheduler start skipped: {e}")

    # 6) Keep process running if scheduler/metrics are active
    with graceful_shutdown(orchestrator=orchestrator, scheduler=scheduler) as term:
        if scheduler is None and not getattr(settings, "enable_metrics", False):
            # No long-running services were started; exit after basic wiring
            print("[app] No scheduler/metrics; exiting after wiring check.")
            return
        print("[app] Running... Ctrl+C to stop")
        while not term["flag"]:
            time.sleep(0.5)


if __name__ == "__main__":
    main()
```

2) Makefile (update or create)
```makefile
# Makefile for TrialSync-ETL

PYTHON ?= python
PIP ?= pip
ENV ?= development

.PHONY: init test run-app run-cli api-smoke

init:
	$(PIP) install -r requirements.txt

# Run unit tests
test:
	pytest -q

# Run the long-lived app (metrics/scheduler). Set ENV=development|test|staging|production
run-app:
	ENVIRONMENT=$(ENV) ENV_FILE=.env.$(ENV) $(PYTHON) -m src.app

# Run the CLI entry (requires CLI module from Issue #19)
# Usage: make run-cli ARGS="run --job-id 1"
run-cli:
	ENVIRONMENT=$(ENV) ENV_FILE=.env.$(ENV) $(PYTHON) -m src.cli.main $(ARGS)

# Smoke test API (if you added src/scripts/api_smoke.py from Quickstart)
api-smoke:
	ENVIRONMENT=$(ENV) ENV_FILE=.env.$(ENV) $(PYTHON) src/scripts/api_smoke.py
```

---
## How to apply (maintainer)
1) Create a new branch, e.g. `feature/wiring-app-metrics-scheduler`.
2) Add the files above:
   - Create `src/app.py` with the provided content
   - Update or create `Makefile` with the provided content
3) Commit and open a PR: "wiring: App runner + Metrics + Scheduler + Makefile".
4) Ensure these dependencies exist in requirements.txt:
   - prometheus-client>=0.20.0
   - apscheduler>=3.10.4 (if using scheduler)
   - psycopg2-binary (or psycopg2) for DB connection
5) Merge after review.

---
## Validation (local)
- Prepare dev env (from Issue #25 and DEV Quickstart):
  - `cp .env.example.development .env.development` and fill non-prod API base URL and test key
  - Ensure Postgres is running and schema + fixes applied
- Run app (metrics + scheduler):
  ```bash
  make init
  make ENV=development run-app
  ```
  - You should see: "Metrics server started on :9090" (if enabled) and "Scheduler started" (if available)
  - Visit http://localhost:9090/metrics to confirm Prometheus endpoint
- Run CLI (requires CLI module from Issue #19):
  ```bash
  make ENV=development run-cli ARGS="run --job-id 1 --dry-run"
  ```

---
## Notes
- src.app uses `load_settings` and `assert_safe_nonprod` from Issue #25; make sure those files are merged (src/config/env.py, src/config/preflight.py).
- If the SchedulerService isn’t merged yet (Issue #18), src.app will skip starting the scheduler and only run metrics (if enabled).
- This wiring is minimal by design; you can extend it to include structured logging (structlog), graceful DB pool shutdown, and richer health checks.

---
## Follow-ups (optional)
- Add a basic health endpoint (e.g., http.server) on a port for readiness/liveness checks
- Integrate Prometheus metrics with job execution hooks for detailed metrics
- Add Dockerfile and docker-compose for local stack
```

**Comment by chrisprader** [comment]
✅ **COMPLETE** - All components implemented and integrated

**Completed:**
- ✅ Web server (`src/web/server.py`)
- ✅ FastAPI application (`src/web/api.py`)
- ✅ Scheduler service (`src/cli/scheduler.py`)
- ✅ Makefile created (see #28)
- ✅ Metrics endpoint (`GET /metrics`)
- ✅ **Metrics fully integrated into execution flow** (see #20)

**All functionality working:**
- Web UI for job management
- Scheduler service with cron support
- Metrics collection and exposure
- Makefile targets for common operations

The application is fully functional with all components integrated. Ready to close.
